docs/enhancements/0003-module-nmc-responsibilty-refactoring.md

```diff
@@ -0,0 +1,110 @@
+
+## Non-Goals
+
+Do not change any other functionality of the two operators, besides their decision making that is described above
+
+## Design
+
+### Module-NMC controller decision-making flow 
+
+The flow takes into account both Module with Version field defined (ordered upgrade) and without Version field defined (un-ordered upgrade)
+Module-NMC does not take into account the current state of the Node (Ready/NotReady/Schedulable/etc'). It just defines if the kernel module should
+be loaded on the node or not based on whether there is a KernelMapping for the current node's kernel and on the labels of the node. All the rest of the decisions
+will be taken by NMC reconciler, which has a much better view of Node's current state and kernel module's current state
+
+1. Found out all the nodes targeted by the Module regardless of node's status, based on the node selector field of the Module
+2. If no suitable KernelMapping for the Node's kernel - do nothing
+3. If there is a suitable KernelMapping and Version field missing in Module (not an ordered upgrade) - update the spec
+4. If there is a suitable KernelMapping, Version field is present in the Module, module loader version label is on the node and
+   its value is equal to the Version - update the spec
+5. If there is a suitable KernelMapping, Version field is present in the Module, module loader version label is on the node and 
+   its value is not equal to Module's version (meaning old version) - do nothing
+6. If there is a suitable KernelMapping, Version field is present in the Module, is missing on the node (meaning kernel module should
+   not be running on the node)  - delete the spec
+
+In this implementation, Module-NMC does not need to delete the spec, but in the 2 following cases:
+1. during ordered upgrade (see point 6 above)
+2. Module is deleted, and so the kernel module should be unloaded
+
+```mermaid
+flowchart TD
+Module[KMM Module]-->|Reconcile| MNC[Module-NMC controller]
+MNC-->|get nodes based on node selector| J1((.))
+J1-->|no KernelMapping for node's kernel| Done[Done]
+J1-->|found KernelMapping for node's kernel| J2((.))
+    J2-->|Version missing in Module| US[Update NMC Spec]
+    J2-->|Version present in Module| J3((.))
+        J3-->|module loader version label equals Version| US
+        J3-->|module loader version label not equal Version| Done
+        J3-->|module loader version label missing| DS[Delete NMC Spec]
+```
+
+
+### NMC controller decision-making flow
+
+NMC takes into account the NMC spec, status, node's status and node's ready timestamp to make decision whether to run worker pods, and whether to run unload or load
+worker pod
+
+1. If Node is not Ready/Schedulable - do nothing
+2. If NMC's status is missing and Node's kernel version equal to NMC's spec kernel version - run worker load pod
+3. If NMC's spec is missing, NMC's status is present and NMC's status kernel version equal to Node's kernel version - run worker unload pod
```

Comment:
Don't we assume at this point that `Module-NMC` has only inserted spec's entries in which the kernelVersion of the kernelMapping matches the node's kernel?

Reply:
NMC controller should not assume anything and should always make sure that it won't run a pod that will fail